### ISSUE TITLE
Suggestion: `yield_` should take a mutable reference

### ISSUE THREAD

**Issue by thomaseizinger** [whatisaphone/genawaiter#15]
Currently, `yield_` takes `&self`: https://github.com/whatisaphone/genawaiter/blob/master/src/core.rs#L131

I would like to start a discussion about changing it to `&mut self`. We ran into a bug in our codebase where we shared a reference of the `Co` object across several futures. Even though we always directly called `.await` on the future returned by `yield_`, we still experienced the panic that tells you to do otherwise.

Turns out that sharing that reference was the problem! I assume what happened was that the future of `yield_` did not completely finish before we called `yield_`a second time from a different future (we `join`ed them together so the executed concurrently).

I think, changing the function to take `&mut self` would solve this because the returned future is already tagged with the lifetime, meaning the Rust compiler should complain about calling `yield` a 2nd time before the first future is dropped.

If I am not mistaken, this might actually completely remove the possibility of misusing the API as in calling `yield` a 2nd time before awaiting it.

What do you think?

**Comment by whatisaphone** [comment]
Can you boil down the problem you were having to a test case I can run? I'm not sure I fully understand the problem.

**Comment by thomaseizinger** [comment]
> Can you boil down the problem you were having to a test case I can run? I'm not sure I fully understand the problem.

Not really I am afraid as it only happened under a bigger load but in pseudo-code what we did was:

1. Instantiate a generator
2. Use `future::try_join` to execute two futures concurrently with both of them getting a reference to `Co`
3. In a loop, call `async_resume().await` and process the events yielded from the generator.

You can see the diff that fixes the bug here: https://github.com/comit-network/comit-rs/pull/2142/commits/8f65871b1b63ca4dd781961e1bfc086046738716#diff-4d0881870d28c4139a6b51ce05586a56L70-L82

There is quite some domain-noise around I am afraid but maybe you get the idea :)

I think the problem was that even though we called `.await` on all `yield_` calls right away, the fact that we shared a reference to `Co` in a `try_join` future made it possible that the panic was triggered. I assume this is because calling `await` is not an atomic operation, you just poll the future. Because we used `try_join`, it was possible for a 2nd code path to call `yield_` without the first one being consumed by the outer call to `async_resume`.

Regardless of the exact behaviour that triggered it, I believe that if we already know that something must not happen, it would be good to make it a compile-error. Mutable references are a great way of ensuring mutual-exclusivity at the type level!
I didn't dig too far but it also seems that there is some unsafe code involved within the implementation of `AirLock` that is only there because of lack of a mutable reference. Thinking about it, this might even be a soundness issue that is only guarded against by the panic and might completely go away if we just used mutable references?

I am happy to explore this idea more and update my PR if you would be interested in merging it :)
But before I do that, could you share some of the motivation behind not using mutable references in the first place? Was it a conscious design decision?

**Comment by whatisaphone** [comment]
Thanks for the link. It looks like you followed genawaiter's docs correctly. Using that commit as a guide I was able to put a repro together:

<details>

```rust
use futures::{executor::block_on, future::join};
use genawaiter::sync::{Co, Gen};

async fn odd_numbers_less_than_ten(co: &Co<i32>) {
    for n in (1..).step_by(2).take_while(|&n| n < 10) {
        co.yield_(n).await;
    }
}

#[test]
fn test_basic() {
    let mut gen = Gen::new(|co| async move {
        join(
            odd_numbers_less_than_ten(&co),
            odd_numbers_less_than_ten(&co),
        ).await
    });
    block_on(async {
        gen.async_resume().await;
    });
}
```

</details>

It's unfortunate that this doesn't work as expected. I think it would still work with certain combinators like `select`, but in the general case it's probably too much of a footgun. I'm surprised your issue was intermittent given how simple the repro is.

About changes to this library: I think the _ideal_ path is to change the type of `co` to `&mut Co<..>`, instead of passing it by value—but when I originally attempted this I couldn't make the compiler happy. 

```rust
// i.e., change this
async fn my_producer(co: Co<u8>) {}
// to this
async fn my_producer(co: &mut Co<u8>) {}
// and change `Gen::new` to match
```

Given that you've already dug into the library a bit (which I appreciate!), you might be interested in reading [the original post I made on the users forum](https://users.rust-lang.org/t/hrtb-on-multiple-generics/34255) a few months ago. I'd be thrilled if you could make more progress than I did. Either way, I'm with you that we should make `yield_` take `&mut self`. Unfortunatley it's a breaking change and (if the above isn't solved) adds annotation burden :/

btw, thanks for introducing me to `trybuild` in that PR, it looks like a great lib. It puts my compile-fail doctests to shame :)

**Comment by thomaseizinger** [comment]
> Thanks for the link. It looks like you followed genawaiter's docs correctly. Using that commit as a guide I was able to put a repro together:
> 
> It's unfortunate that this doesn't work as expected. I think it would still work with certain combinators like `select`, but in the general case it's probably too much of a footgun. I'm surprised your issue was intermittent given how simple the repro is.
> 
> About changes to this library: I think the _ideal_ path is to change the type of `co` to `&mut Co<..>`, instead of passing it by value—but when I originally attempted this I couldn't make the compiler happy.
> 
> ```rust
> // i.e., change this
> async fn my_producer(co: Co<u8>) {}
> // to this
> async fn my_producer(co: &mut Co<u8>) {}
> // and change `Gen::new` to match
> ```
> 
> Given that you've already dug into the library a bit (which I appreciate!), you might be interested in reading [the original post I made on the users forum](https://users.rust-lang.org/t/hrtb-on-multiple-generics/34255) a few months ago. I'd be thrilled if you could make more progress than I did. Either way, I'm with you that we should make `yield_` take `&mut self`. Unfortunatley it's a breaking change and (if the above isn't solved) adds annotation burden :/
> 
> btw, thanks for introducing me to `trybuild` in that PR, it looks like a great lib. It puts my compile-fail doctests to shame :)

Thanks for the input!

I read through the forum post but couldn't really follow the problem to be honest. At least I couldn't map it to the domain of genawaiter in my head :)

[The PR](https://github.com/whatisaphone/genawaiter/pull/16) I put up changes `Co::yield_` to take `&mut self` and passes all the tests, so I am not quite sure, what exactly would not be working :(

Could you push additional tests (to master preferable, so I can just rebase my branch) that verify certain usage patterns which currently work but would be broken with `Co` taking `&mut self`?

**Comment by thomaseizinger** [comment]
Ah now I get you!

> I think the ideal path is to change the type of co to &mut Co<..>, instead of passing it by value

I will try to make that work and come back to you :)

**Comment by thomaseizinger** [comment]
Hmm, I played with this a bit but HRTB are not really my field of expertise in Rust, been avoiding those as much as possible so far :D 

To be honest, I cannot get even past the `new` function of `Co`. How are you mean to compute something (producer) with only a

**Comment by whatisaphone** [comment]
Yeah, sorry for the confusion! Your PR is great. What I'm getting at is there's an approach that seems better to me (this library's white whale, you might say), but I'm not sure if it's possible to express in the current version of Rust. I thought I'd pose the question since you've dug into the library a bit. If you don't have any luck, I'll merge your PR as is. Note: don't feel obligated if you just think I'm spouting nonsense here 🙂

---

>What do you think would make a solution that uses &mut Co superior to passing an owned version of Co?

The `Co` semantically contains a borrow of the airlock. The airlock is semantically owned by the generator. Right now, the annotations don't match the semantics, and so the compiler allows the `Co` to live beyond the life of the future. Because of this, the library has some cruft to ensure there is no unsafety if the user does certain silly things with the `Co` object. If the `Co` could be changed to `&mut Co`, we could get rid of the cruft (some panics, and likely some allocations and unsafe code).

<details>
<summary>For example, this code is silly, but even silly (but safe) code must not result in memory unsafety</summary>

```rust
let co_escaped: Cell<Option<Co<i32>>> = Cell::new(None);
let mut gen = Gen::new(|co| async {
    co_escaped.set(Some(co));
});
gen.resume();
co_escaped.take().unwrap().yield_(10);
```

</details>

---

>At least I couldn't map it to the domain of genawaiter in my head :)

Yeah sorry, it's pretty obfuscated isn't it, forget about the forum post. This might be a better way of explaining – Here's the cut-down version of the lib I used for playing with lifetimes:

<details>
<summary>Details</summary>

```rust
use futures::future;
use std::{cell::Cell, future::Future, marker::PhantomData};

struct Gen<Y, R, F>(PhantomData<(Y, R, F)>);

impl<Y, R, F: Future> Gen<Y, R, F> {
    pub fn new(producer: impl FnOnce(Co<Y, R>) -> F) -> Self {
        unimplemented!()
    }
}

struct Co<Y, R = ()>(PhantomData<(Y, R)>);

impl<Y> Co<Y, ()> {
    pub fn yield_(&self, value: Y) -> impl Future<Output = ()> {
        future::ready(())
    }
}

#[test]
fn test() {
    let co_escaped: Cell<Option<Co<i32>>> = Cell::new(None);
    let gen = Gen::new(|co| async move {
        co.yield_(10).await;
        // co_escaped.set(Some(co));
    });
}
```

</details>

The goal is to change `Co` to `&mut Co` like so:

```diff
 impl<Y, R, F: Future> Gen<Y, R, F> {
-    pub fn new(producer: impl FnOnce(Co<Y, R>) -> F) -> Self {
+    pub fn new(producer: impl FnOnce(&mut Co<Y, R>) -> F) -> Self {

 impl<Y> Co<Y, ()> {
-    pub fn yield_(&self, value: Y) -> impl Future<Output = ()> {
+    pub fn yield_(&mut self, value: Y) -> impl Future<Output = ()> {

 fn test() {
-    let co_escaped: Cell<Option<Co<i32>>> = Cell::new(None);
+    let co_escaped: Cell<Option<&mut Co<i32>>> = Cell::new(None);
```

… and then figure out how to make the code cleanly compile. And the commented line with `co_escaped.set` should cause a lifetime error when uncommented (this is the same silly code as in my example earlier).

If you can figure it out in this toy example, I'm happy to do the actual work of porting it to the real lib. If not, I may have to accept that it's not possible. In that case, I'll merge your PR (pending closer review)

**Comment by thomaseizinger** [comment]
Very interesting problem! Thanks for writing all this up :)

I had an idea but it is kind of ugly and I am not sure if it fully solves the problem:

<details>
use futures::{future, FutureExt};
use std::{cell::Cell, future::Future, marker::PhantomData};
use std::pin::Pin;
use futures::future::BoxFuture;

struct Gen<Y, R, F>(PhantomData<(Y, R, F)>);

trait Producer<Y, R, E> {
    fn new(co: Pin<Box<Co<u32, ()>>>, env: E) -> Self;
    fn produce(&mut self) -> BoxFuture<'_, R>;
}

struct MyProducer<Y> {
    co: Pin<Box<Co<Y, ()>>>
}

impl Producer<u32, (), ()> for MyProducer<u32> {
    fn new(co: Pin<Box<Co<u32, ()>>>, env: ()) -> Self {
        Self {
            co
        }
    }

    fn produce(&mut self) -> BoxFuture<'_, ()> {
        async move {
            // co.as_mut().yield_(10).await;
            self.co.as_mut().yield_(10).await;
            // co_escaped.set(Some(co));
        }.boxed()
    }
}

impl<Y, R, F: Future> Gen<Y, R, F> {
    pub fn new<E, P: Producer<Y, R, E>>(env: E) -> Self {
    // pub fn new(producer: impl FnOnce(Co<Y, R>) -> F) -> Self {
        let co = Box::pin(Co(PhantomData));

        let mut producer = P::new(co, env);

        let f = producer.produce();

        unimplemented!()
    }
}

struct Co<Y, R = ()>(PhantomData<(Y, R)>);

impl<Y> Co<Y, ()> {
    pub fn yield_(&mut self, value: Y) -> impl Future<Output = ()> {
        future::ready(())
    }
}

#[test]
fn test() {
    let co_escaped: Cell<Option<Co<i32>>> = Cell::new(None);
    let gen = Gen::<_, _, BoxFuture<'_, ()>>::new::<(), MyProducer<u32>>(());
}
</details>

Basically, we get rid of the closure so we have full control over the interface that the user has access to.
We then force the user to implement a trait that provides a `new` function which _we_ are going to use to construct the instance. We also allow the user to provide some `env` data that we are going to pass to them on construction time.

I am not sure if the `Pin<Box<Co>>` is necessary but maybe you get the idea. If the user wants to take the `Co` object and move it somewhere else inside `new` - fine, then they will not have it available when we call `produce` and whatever they do in their cannot use `Co`.

I don't think a solution like this is worth the trade-off :)
Closures are really handy and nice to work with.

**Comment by DevinR528** [comment]
[Here's](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b790927598906cd5bbe09488164ae855) a playground link if that's helpful

**Comment by whatisaphone** [comment]
Thanks for taking a stab at it. That's an interesting direction, but I agree with you that it's not worth the tradeoff. I'll go ahead and merge your original PR. Thanks for your contribution!

**Comment by thomaseizinger** [comment]
Thanks a lot, I consider this issue to be resolved then :)
